onnxruntime/core/common/profiler.cc

```diff
@@ -7,13 +7,27 @@ namespace onnxruntime {
 namespace profiling {
 using namespace std::chrono;
 
+#ifdef ENABLE_STATIC_PROFILER_INSTANCE
+Profiler* Profiler::instance_ = nullptr;
+
+profiling::Profiler::~Profiler() {
+  instance_ = nullptr;
+}
+#else
+profiling::Profiler::~Profiler() {}
+#endif
+
 ::onnxruntime::TimePoint profiling::Profiler::StartTime() const {
   return std::chrono::high_resolution_clock::now();
 }
 
 void Profiler::Initialize(const logging::Logger* session_logger) {
   ORT_ENFORCE(session_logger != nullptr);
   session_logger_ = session_logger;
+#ifdef ENABLE_STATIC_PROFILER_INSTANCE
+  ORT_ENFORCE(instance_ == nullptr, "Static profiler instance only works with single session");
+  instance_ = this;
+#endif
 }
```

Comment:
This is a little confusing as I would have expected an actual static instance to be in this file. Is the usage that you create a profiler instance externally and this is just a pointer to that? Setting a static member via a call to a non-static method is also a bit unexpected. Is there a more standard approach we could use?

Reply:
In current ORT design, profiler goes with inference session. Since it's possible to have multiple inference sessions, profiler by definition is not singleton thus cannot use that design pattern. This is why I add another macro to special case the access only when there's one inference session. This is needed to access profiler in code that have no access to session (like inside TVM extern functions)

---
In reply to: [313176458](https://github.com/microsoft/onnxruntime/pull/881#discussion_r313176458) [](ancestors = 313176458)